Sources/NIO/PendingDatagramWritesManager.swift

```diff
@@ -101,19 +113,41 @@ private func doPendingDatagramWriteVectorOperation(pending: PendingDatagramWrite
         p.data.withUnsafeReadableBytesWithStorageManagement { ptr, storageRef in
             storageRefs[c] = storageRef.retain()
             let addressLen = p.copySocketAddress(addresses.baseAddress! + c)
-            iovecs[c] = iovec(iov_base: UnsafeMutableRawPointer(mutating: ptr.baseAddress!), iov_len: numericCast(toWriteForThisBuffer))
+
+            let buffer: UnsafeMutablePointer<IOVector> = iovecs.baseAddress! + c
+            let address: UnsafeMutablePointer<sockaddr_storage> =
+                addresses.baseAddress! + c
+
+#if os(Windows)
+            iovecs[c] =
+                IOVector(len: numericCast(toWriteForThisBuffer),
+                         buf: UnsafeMutableRawPointer(mutating: ptr.baseAddress!).assumingMemoryBound(to: CHAR.self))
+#else
+            iovecs[c] = IOVector(iov_base: UnsafeMutableRawPointer(mutating: ptr.baseAddress!), iov_len: numericCast(toWriteForThisBuffer))
+#endif
 
             var controlBytes = UnsafeOutboundControlBytes(controlBytes: controlMessageStorage[c])
             controlBytes.appendExplicitCongestionState(metadata: p.metadata, protocolFamily: p.address.protocol)
             let controlMessageBytePointer = controlBytes.validControlBytes
 
-            let msg = msghdr(msg_name: addresses.baseAddress! + c,
+#if os(Windows)
+            let msg = address.withMemoryRebound(to: SOCKADDR.self, capacity: 1) {
```

Comment:
This `withMemoryRebound` dance, both here and above, is not right. We are escaping the rebound pointer, which violates Swift's pointer aliasing rules. Is `SOCKADDR` a meaningfully different type, or is it the same as `sockaddr_storage` in all meaningful ways?

Reply:
They are different.  The `SOCKADDR_STORAGE` is opaque content, and will need to be cast to the `SOCKADDR`.  From the documentation:

> A WSK application typically does not directly access any of the members of the SOCKADDR_STORAGE structure except for the ss_family member. Instead, a pointer to a SOCKADDR_STORAGE structure is normally cast to a pointer to the specific SOCKADDR structure type that corresponds to a particular address family.

Reply:
Ugh this really awkward, this is extremely hard to fix. I think we need to write a thunk version of the function we want to call that takes a structure that isn't `WSAMSG` but is _basically_ a `WSAMSG` that takes a `SOCKADDR_STORAGE` and not a `SOCKADDR`. Swift pointer aliasing forbids us from escaping the temporarily-rebound pointer to the `SOCKADDR`, so we have to rebind at a higher scope around this whole function which is also extremely difficult to do.

Thunking through C is gonna be necessary.

Reply:
I don't understand what you mean by a thunk here that takes that - this is being used to construct an array of elements of these types, I don't see how to thunk the function call without sinking all of this into C.

Reply:
I think we only need to thunk the operation at the bottom of the call stack (into libc) to take the function type we want, no? Then the changes will ripple up through Swift and allow us to remove the rebind here.

Reply:
You are missing something. :smile:

We cannot construct a `WSAMSG` here, in this way, because to do so requires us to escape a rebound pointer and violate aliasing rules, which we may not do.

_However_, `MMsgHdr` on Windows is a type _we_ defined, and so _we_ control it. We can therefore redefine it to store, instead of a `WSAMSG`, a type almost exactly the same as `WSAMSG` but with the socket address pointer type we want to use from Swift. The C code can then cast the buffer to the appropriate type internally, where this cast is "safe".